compiler/src/dotty/tools/dotc/CompilationUnit.scala

```diff
@@ -115,7 +115,7 @@ object CompilationUnit {
   /** Make a compilation unit for top class `clsd` with the contents of the `unpickled` tree */
   def apply(clsd: ClassDenotation, unpickled: Tree, forceTrees: Boolean)(using Context): CompilationUnit =
     val file = clsd.symbol.associatedFile.nn
-    apply(SourceFile(file, Array.empty[Char]), unpickled, forceTrees)
+    apply(ctx.getEmptySource(file), unpickled, forceTrees)
```

Comment:
Is this still useful?

Reply:
with this basically all SourceFile constructions will be cached

Reply:
except virtual ones, but they are reference equality

Reply:
I removed it